electron: build from source

yu-re-ka opened the pull request:
## Description of changes

Replacement of #207766 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

amarshall reviewed (commented):
Fantastic work, really looking forward to this being merged! I’ve done:

- A cursory review (in particular, mostly skipped the update script); caveat being I don’t have a ton of context on the Chromium build process; none of my comments should be considered blocking in any way
- Built on electron-src.electron_25 on x86_64-linux
- Temporarily changed bitwarden to use electron_25, built, and done basic testing—everything seemed good

yu-re-ka commented on pkgs/applications/networking/browsers/chromium/default.nix:
```diff
@@ -28,14 +29,12 @@ let
-         (lib.versionAtLeast ungoogled-version min-version)
-         "chromium: ungoogled version ${ungoogled-version} is newer than a conditional bounded at ${min-version}. You can safely delete it."
+         (lib.versionAtLeast min-supported-version min-version)
+         "chromium: min-supported-version is newer than a conditional bounded at ${min-version}"
```
Of the supported flavors (chromium, chromium-beta, ungoogled-chrome), the ungoogled-chrome channel would always be the most behind for practical reasons (porting the patches takes time).
So once ungoogled-chrome had advanced past a certain version, the expressions could be cleaned up to remove this version.

Now it is most likely the oldest supported electron version, because that is usually 4 versions behind stable (compared to ungoogled-chrome's 1-2).
I will think about implementing this again with some logic that selects the chromium version of the oldest electron version that we support.

yu-re-ka commented on pkgs/development/tools/electron/common.nix:
```diff
@@ -0,0 +1,180 @@
+    echo 'generate_location_tags = true' >> build/config/gclient_args.gni
+
+    echo 'LASTCHANGE=${info.deps."src".rev}-refs/heads/master@{#0}'        > build/util/LASTCHANGE
+    echo '1555555555'                                                      > build/util/LASTCHANGE.committime
```
Good point, but then we also need to add `runHook postUnpack`, since otherwise the hook that sets this variable is never run

yu-re-ka commented on pkgs/development/tools/electron/common.nix:
```diff
@@ -0,0 +1,180 @@
+
+    (
+      cd ..
+      PATH=$PATH:${lib.makeBinPath [ jq git ]}
```
Adding packages to nativeBuildInputs is usually preferred, because this will not work properly when cross-compiling.

yu-re-ka commented:
- Addressed review comments (in individual commits for easier review)
- Rebased on latest master (since staging-next was merged)
- Updated info.json for new electron versions 27.0.0-alpha.6 -> 27.0.0-beta.2, 26.2.0 -> 26.2.1, 25.8.0 -> 25.8.1
- pushed new versions to the binary cache

yu-re-ka commented:
I am willing to press the merge button myself in a week or so, if nobody vetos ;)

yorickvP commented:
The `.headers` attribute is missing on electron_27, at least. Could we generate that?

yu-re-ka commented:
- added node native API headers derivation, and exposed via `electron.headers`
- Updated info.json for new electron versions: 25.8.1 -> 25.8.2, 26.2.1 -> 26.2.2, 27.0.0-beta.2 -> 27.0.0-beta.4
- Rebased on latest master
- Tested cross-compilation: #256733
- pushed new versions to the binary cache

yu-re-ka commented:
CC @travisbhartwell @manveru @prusnak

yuyuyureka committed:
electron-source.electron_26: init at 26.2.3

yuyuyureka committed:
chromium: changes required for electron

yuyuyureka committed:
electron_{25,26}: default to source-built version on linux

SuperSandro2000 commented:
You rock pushing through with this! ❤️
Remind to buy you a Mate next time we meet.
